storage/index/types.go

```diff
@@ -183,42 +179,306 @@ type BlockTickResult struct {
 			// in the slice is a special case because we are always starting a new group at that point.
 			if i == 0 {
 				continue
 			}
-			fn(w[startIdx].BlockStart, w[startIdx:i])
+			b.entries = allEntries[startIdx:i]
+			b.docs = allDocs[startIdx:i]
+			fn(prevLastBlockStart, b)
 			startIdx = i
 		}
 	}
+
 	// spill over
-	if startIdx < len(w) {
-		fn(w[startIdx].BlockStart, w[startIdx:])
+	if startIdx < len(allEntries) {
+		b.entries = allEntries[startIdx:]
+		b.docs = allDocs[startIdx:]
+		fn(lastBlockStart.ToTime(), b)
+	}
+}
+
+func (b *WriteBatch) numPending() int {
+	numUnmarked := 0
+	for i := range b.entries {
+		if b.entries[i].OnIndexSeries == nil {
+			break
+		}
+		numUnmarked++
 	}
+	return numUnmarked
+}
+
+// PendingDocs returns all the docs in this batch that are unmarked.
+func (b *WriteBatch) PendingDocs() []doc.Document {
+	b.SortByUnmarkedAndIndexBlockStart() // Ensure sorted by unmarked first
+	return b.docs[:b.numPending()]
+}
+
+// PendingEntries returns all the entries in this batch that are unmarked.
+func (b *WriteBatch) PendingEntries() []WriteBatchEntry {
+	b.SortByUnmarkedAndIndexBlockStart() // Ensure sorted by unmarked first
+	return b.entries[:b.numPending()]
+}
+
+// NumErrs returns the number of errors encountered by the batch.
+func (b *WriteBatch) NumErrs() int {
+	errs := 0
+	for _, entry := range b.entries {
+		if entry.result.Err != nil {
+			errs++
+		}
+	}
+	return errs
+}
+
+// Reset resets the batch for use.
+func (b *WriteBatch) Reset() {
+	// Memset optimizations
+	var entryZeroed WriteBatchEntry
+	for i := range b.entries {
+		b.entries[i] = entryZeroed
+	}
+	b.entries = b.entries[:0]
+	var docZeroed doc.Document
+	for i := range b.docs {
+		b.docs[i] = docZeroed
+	}
+	b.docs = b.docs[:0]
+}
+
+// SortByUnmarkedAndIndexBlockStart sorts the batch by unmarked first and then
+// by index block start time.
+func (b *WriteBatch) SortByUnmarkedAndIndexBlockStart() {
+	b.sortBy = writeBatchSortByUnmarkedAndBlockStart
+	sort.Stable(b)
+}
+
+// SortByEnqueued sorts the entries and documents back to the sort order they
+// were enqueued as.
+func (b *WriteBatch) SortByEnqueued() {
+	b.sortBy = writeBatchSortByEnqueued
+	sort.Stable(b)
+}
+
+// MarkUnmarkedEntriesSuccess marks all unmarked entries as success.
+func (b *WriteBatch) MarkUnmarkedEntriesSuccess() {
+	for idx := range b.entries {
+		if b.entries[idx].OnIndexSeries != nil {
+			blockStart := b.entries[idx].indexBlockStart(b.opts.IndexBlockSize)
+			b.entries[idx].OnIndexSeries.OnIndexSuccess(blockStart)
+			b.entries[idx].OnIndexSeries.OnIndexFinalize(blockStart)
+			b.entries[idx].OnIndexSeries = nil
+			b.entries[idx].result = WriteBatchEntryResult{Err: nil}
+		}
+	}
+}
+
+// MarkUnmarkedEntriesError marks all unmarked entries as error.
+func (b *WriteBatch) MarkUnmarkedEntriesError(err error) {
+	for idx := range b.entries {
+		if b.entries[idx].OnIndexSeries != nil {
```

Comment:
hm, can you reuse the method below, i.e.

``` 
func (b *WriteBatch) MarkUnmarkedEntriesError(err error) {
  for idx := range b.entries {
    b.MarkUnmarkedEntryError(err, idx)
  }
}
```

Reply:
Sure thing, sounds good.